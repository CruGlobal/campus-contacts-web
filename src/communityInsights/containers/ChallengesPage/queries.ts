import gql from 'graphql-tag';

const GET_IMPACT_CHALLENGES = gql`
    query impactReportStepsCount($id: ID!) {
        community(id: $id) {
            impactReport {
                stepsCount
            }
        }
    }
`;

const GET_CHALLENGES = gql`
    query communityChallenges(
        $id: ID!
        $first: Int
        $last: Int
        $after: String
        $before: String
        $sortBy: [CommunityChallengesSortEnum!]
    ) {
        community(id: $id) {
            communityChallenges(
                first: $first
                last: $last
                after: $after
                before: $before
                sortBy: $sortBy
            ) {
                nodes {
                    id
                    title
                    acceptedCount
                    completedCount
                    createdAt
                    endDate
                }
                pageInfo {
                    hasNextPage
                    startCursor
                    endCursor
                    hasPreviousPage
                }
            }
        }
    }
`;

export { GET_CHALLENGES, GET_IMPACT_CHALLENGES };
