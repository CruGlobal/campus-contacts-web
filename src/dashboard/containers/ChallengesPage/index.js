import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import AppContext from '../../appContext';
import Table from '../../components/Table';
import ImpactInfo from '../../components/ImpactInfo';

import {
    GET_IMPACT_CHALLENGES,
    GET_CHALLENGES,
    GET_GLOBAL_CHALLENGES,
} from './queries';

const ChallengesPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    const mapRows = data => {
        const percentage = (a, b) => {
            const value = b === 0 ? 0 : Math.floor((a / b) * 100);
            return `${a} (${value}%)`;
        };

        const duration = (start, end) => {
            const startDate = moment(start).startOf('day');
            const endDate = moment(end).startOf('day');
            const days = moment.duration(endDate.diff(startDate)).asDays();
            return days === 1 ? '1 Day' : `${days} Days`;
        };

        const range = (start, end) => {
            const format = 'MM/DD/YYYY';
            const startDate = moment(start).format(format);
            const endDate = moment(end).format(format);
            return `${startDate} - ${endDate}`;
        };

        const rows = [];
        data.globalCommunityChallenges.nodes.forEach(
            ({
                id,
                title,
                acceptedCount,
                completedCount,
                createdAt,
                endDate,
            }) => {
                const cells = [];
                cells.push(id);
                cells.push(title);
                cells.push(acceptedCount);
                cells.push(percentage(completedCount, acceptedCount));
                cells.push(duration(createdAt, endDate));
                cells.push(range(createdAt, endDate));
                rows.push(cells);
            },
        );
        return rows;
    };

    const mapPage = data => {
        return data.globalCommunityChallenges.pageInfo;
    };

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_CHALLENGES}
                text={report =>
                    t('challenges.total', {
                        count:
                            report.communityReport.impactReport.challengesCount,
                    })
                }
                variables={{ organizationId: orgId }}
            />
            <Card
                title={t('challenges.title')}
                subtitle={t('challenges.subTitle')}
                noPadding={true}
            >
                <Table
                    query={GET_GLOBAL_CHALLENGES}
                    variables={{ id: orgId }}
                    headers={[
                        '',
                        '',
                        t('challenges.join'),
                        t('challenges.completed'),
                        t('challenges.length'),
                        t('challenges.dates'),
                    ]}
                    mapRows={mapRows}
                    mapPage={mapPage}
                />
            </Card>
        </div>
    );
};

export default ChallengesPage;

ChallengesPage.propTypes = {};
