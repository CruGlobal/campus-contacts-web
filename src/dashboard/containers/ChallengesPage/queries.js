import gql from 'graphql-tag';

const GET_IMPACT_CHALLENGES = gql`
    query communityReport {
        communityReport {
            impactReport {
                interactionsCount
                interactionsReceiversCount
                challengesCount
            }
        }
    }
`;

const GET_CHALLENGES = gql`
    query globalCommunityChallenges($id: ID!, $first: Int, $after: String) {
        community(id: $id) {
            communityChallenges(first: $first, after: $after) {
                nodes {
                    id
                    title
                    acceptedCount
                    completedCount
                    createdAt
                    endDate
                }
                pageInfo {
                    hasNextPage
                    startCursor
                    endCursor
                    hasPreviousPage
                }
            }
        }
    }
`;

export { GET_CHALLENGES, GET_IMPACT_CHALLENGES };
