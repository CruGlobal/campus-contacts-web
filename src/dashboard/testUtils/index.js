import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
