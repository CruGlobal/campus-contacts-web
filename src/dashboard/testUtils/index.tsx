import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
