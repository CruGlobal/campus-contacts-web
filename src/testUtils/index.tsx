import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
