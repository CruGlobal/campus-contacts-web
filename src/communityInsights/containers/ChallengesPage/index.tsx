import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import { AppContext } from '../../../appContext';
import Table from '../../components/Table';
import ImpactInfo from '../../components/ImpactInfo';

import { GET_IMPACT_CHALLENGES, GET_CHALLENGES } from './queries';
// GET_IMPACT_CHALLENGES INTERFACES
import { impactReportStepsCount } from './__generated__/impactReportStepsCount';
// GET_CHALLENGES INTERFACES
import {
    communityChallenges,
    communityChallenges_community_communityChallenges_nodes as communityChallengesNodes,
} from './__generated__/communityChallenges';

const ChallengesPage = () => {
    const { t } = useTranslation('insights');
    const { orgId } = useContext(AppContext);

    const mapRows = (data: communityChallenges) => {
        const percentage = (a: number, b: number) => {
            const value = b === 0 ? 0 : Math.floor((a / b) * 100);
            return `${a} (${value}%)`;
        };

        const duration = (start: string, end: string) => {
            const startDate = moment(start).startOf('day');
            const endDate = moment(end).startOf('day');
            const days = Math.floor(
                moment.duration(endDate.diff(startDate)).asDays(),
            );
            return days === 1 ? '1 Day' : `${days} Days`;
        };

        const range = (start: string, end: string) => {
            const format = 'MM/DD/YYYY';
            const startDate = moment(start).format(format);
            const endDate = moment(end).format(format);
            return `${startDate} - ${endDate}`;
        };

        return data.community.communityChallenges.nodes.map(
            ({
                id,
                title,
                acceptedCount,
                completedCount,
                createdAt,
                endDate,
            }: communityChallengesNodes) => [
                id,
                title,
                acceptedCount,
                percentage(completedCount, acceptedCount),
                duration(createdAt, endDate),
                range(createdAt, endDate),
            ],
        );
    };

    const mapPage = (data: communityChallenges) => {
        return data.community.communityChallenges.pageInfo;
    };

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_CHALLENGES}
                text={(report: impactReportStepsCount) =>
                    t('challenges.total', {
                        count: report.community.impactReport.stepsCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card
                title={t('challenges.title')}
                subtitle={t('challenges.subTitle')}
                noPadding={true}
                noMarginBottom={true}
            >
                <Table
                    query={GET_CHALLENGES}
                    variables={{
                        id: orgId,
                        sortBy: 'createdAt_DESC',
                    }}
                    headers={[
                        '',
                        '',
                        t('challenges.join'),
                        t('challenges.completed'),
                        t('challenges.length'),
                        t('challenges.dates'),
                    ]}
                    mapRows={mapRows}
                    mapPage={mapPage}
                />
            </Card>
        </div>
    );
};

export default ChallengesPage;
