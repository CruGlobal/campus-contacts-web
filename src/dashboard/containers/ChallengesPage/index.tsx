import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';
import moment from 'moment';

import Card from '../../components/Card';
import AppContext from '../../appContext';
import Table from '../../components/Table';
import ImpactInfo from '../../components/ImpactInfo';

import { GET_IMPACT_CHALLENGES, GET_CHALLENGES } from './queries';

const ChallengesPage = () => {
    const { t } = useTranslation('insights');
    const { orgId }: any = useContext(AppContext);

    const mapRows = (data: any) => {
        const percentage = (a: number, b: number) => {
            const value = b === 0 ? 0 : Math.floor((a / b) * 100);
            return `${a} (${value}%)`;
        };

        const duration = (start: Date, end: Date) => {
            const startDate = moment(start).startOf('day');
            const endDate = moment(end).startOf('day');
            const days = Math.floor(
                moment.duration(endDate.diff(startDate)).asDays(),
            );
            return days === 1 ? '1 Day' : `${days} Days`;
        };

        const range = (start: Date, end: Date) => {
            const format = 'MM/DD/YYYY';
            const startDate = moment(start).format(format);
            const endDate = moment(end).format(format);
            return `${startDate} - ${endDate}`;
        };

        return data.community.communityChallenges.nodes.map(
            ({
                id,
                title,
                acceptedCount,
                completedCount,
                createdAt,
                endDate,
            }: any) => [
                id,
                title,
                acceptedCount,
                percentage(completedCount, acceptedCount),
                duration(createdAt, endDate),
                range(createdAt, endDate),
            ],
        );
    };

    const mapPage = (data: any) => {
        return data.community.communityChallenges.pageInfo;
    };

    return (
        <div>
            <ImpactInfo
                query={GET_IMPACT_CHALLENGES}
                text={report =>
                    t('challenges.total', {
                        count: report.community.impactReport.interactionsCount,
                    })
                }
                variables={{ id: orgId }}
            />
            <Card
                title={t('challenges.title')}
                subtitle={t('challenges.subTitle')}
                noPadding={true}
            >
                <Table
                    query={GET_CHALLENGES}
                    variables={{
                        id: orgId,
                        sortBy: 'createdAt_DESC',
                    }}
                    headers={[
                        '',
                        '',
                        t('challenges.join'),
                        t('challenges.completed'),
                        t('challenges.length'),
                        t('challenges.dates'),
                    ]}
                    mapRows={mapRows}
                    mapPage={mapPage}
                />
            </Card>
        </div>
    );
};

export default ChallengesPage;

ChallengesPage.propTypes = {};
