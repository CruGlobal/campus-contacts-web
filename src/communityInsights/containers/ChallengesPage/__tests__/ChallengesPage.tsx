import React from 'react';

import { renderWithContext } from '../../../../testUtils';

import ChallengesPage from '..';

describe('<ChallengesPage />', () => {
    it('should render properly in loading state', () => {
        const { snapshot, unmount } = renderWithContext(<ChallengesPage />, {
            appContext: {
                orgId: '1',
            },
        });
        snapshot();
        unmount();
    });

    // TODO: Add a test with mocking data when GraphQL schema contains above query
});
