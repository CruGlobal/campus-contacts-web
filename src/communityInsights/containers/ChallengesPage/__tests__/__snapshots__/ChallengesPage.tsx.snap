// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChallengesPage /> should render properly after loading 1`] = `
.emotion-0 {
  font-weight: 300;
  font-size: 30px;
  line-height: 38px;
  color: #505256;
  margin-bottom: 0px;
  margin-top: 48px;
  padding-left: 21px;
}

.emotion-36 {
  box-shadow: 0px 2px 12px rgba(0,0,0,0.2);
  background: #ffffff;
  border-radius: 8px;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 43px;
  margin-top: 36px;
}

.emotion-1 {
  font-weight: 300;
  font-size: 36px;
  line-height: 38px;
  color: #505256;
  margin-bottom: 12px;
  margin-top: 0px;
}

.emotion-2 {
  font-size: 14px;
  line-height: 20px;
  font-weight: normal;
  color: #505256;
  margin-top: 0;
  margin-bottom: -40px;
}

.emotion-35 {
  padding: 0;
  margin: 0 -24px;
}

.emotion-25 {
  table-layout: fixed;
  width: 100%;
}

.emotion-9 {
  text-align: left;
}

.emotion-3 {
  font-weight: normal;
  font-size: 14px;
  line-height: 60px;
  color: #B4B6BA;
  padding: 0 15px;
}

.emotion-16 {
  text-align: left;
  height: 64px;
}

.emotion-16:nth-of-type(odd) {
  background-color: #eceef2;
}

.emotion-10 {
  font-weight: normal;
  font-size: 14px;
  line-height: 20px;
  color: #505256;
  padding: 0 15px;
  max-width: 200px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-10:first-of-type {
  padding: 0 25px;
}

.emotion-10:last-child {
  padding: 0 5px;
}

.emotion-34 {
  width: 100%;
  height: 64px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-26 {
  height: 16px;
  width: 16px;
  background-size: contain;
  background-repeat: no-repeat;
  background-image: url(mockFile);
  margin-right: 20px;
}

.emotion-26.active {
  cursor: pointer;
  background-image: url(mockFile);
}

.emotion-32 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-27 {
  width: 32px;
  height: 32px;
  font-size: 14px;
  line-height: 20px;
  color: #B4B6BA;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-27:not(:last-child) {
  margin-right: 4px;
}

.emotion-27.active {
  border-radius: 16px;
  background-color: #3CC8E6;
  color: #ffffff;
}

.emotion-33 {
  height: 16px;
  width: 16px;
  background-size: contain;
  background-repeat: no-repeat;
  background-image: url(mockFile);
  margin-left: 20px;
}

.emotion-33.active {
  cursor: pointer;
  background-image: url(mockFile);
}

<div>
  <div>
    <h1
      class="emotion-0"
    >
      This year, we have completed 41702 challenges.
       
    </h1>
    <div
      class="emotion-36"
    >
      <h1
        class="emotion-1"
      >
        Challenges
      </h1>
      <h2
        class="emotion-2"
      >
        Community challenges joined and completed.
      </h2>
      <div
        class="emotion-35"
      >
        <div>
          <table
            class="emotion-25"
          >
            <thead
              class="emotion-9"
            >
              <tr>
                <th
                  class="emotion-3"
                />
                <th
                  class="emotion-3"
                />
                <th
                  class="emotion-3"
                >
                  Joined
                </th>
                <th
                  class="emotion-3"
                >
                  Completed
                </th>
                <th
                  class="emotion-3"
                >
                  Length
                </th>
                <th
                  class="emotion-3"
                >
                  Dates
                </th>
              </tr>
            </thead>
            <tbody
              class="emotion-24"
            >
              <tr
                class="emotion-16"
              >
                <td
                  class="emotion-10"
                >
                  1
                </td>
                <td
                  class="emotion-10"
                >
                  repellat quisquam recusandae
                </td>
                <td
                  class="emotion-10"
                >
                  11
                </td>
                <td
                  class="emotion-10"
                >
                  12812 (116472%)
                </td>
                <td
                  class="emotion-10"
                >
                  -2543 Days
                </td>
                <td
                  class="emotion-10"
                >
                  01/05/2017 - 01/19/2010
                </td>
              </tr>
              <tr
                class="emotion-16"
              >
                <td
                  class="emotion-10"
                >
                  2
                </td>
                <td
                  class="emotion-10"
                >
                  ratione ut sunt
                </td>
                <td
                  class="emotion-10"
                >
                  39658
                </td>
                <td
                  class="emotion-10"
                >
                  18626 (46%)
                </td>
                <td
                  class="emotion-10"
                >
                  154 Days
                </td>
                <td
                  class="emotion-10"
                >
                  02/27/2016 - 07/31/2016
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="emotion-34"
          >
            <div
              class=" emotion-26"
            />
            <div
              class="emotion-32"
            >
              <div
                class="active emotion-27"
              >
                1
              </div>
              <div
                class=" emotion-27"
              >
                2
              </div>
              <div
                class=" emotion-27"
              >
                3
              </div>
              <div
                class=" emotion-27"
              >
                4
              </div>
              <div
                class=" emotion-27"
              >
                5
              </div>
            </div>
            <div
              class=" emotion-33"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ChallengesPage /> should render properly in loading state 1`] = `
.emotion-0 {
  font-weight: 300;
  font-size: 30px;
  line-height: 38px;
  color: #505256;
  margin-bottom: 0px;
  margin-top: 48px;
  padding-left: 21px;
}

.emotion-5 {
  box-shadow: 0px 2px 12px rgba(0,0,0,0.2);
  background: #ffffff;
  border-radius: 8px;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 43px;
  margin-top: 36px;
}

.emotion-1 {
  font-weight: 300;
  font-size: 36px;
  line-height: 38px;
  color: #505256;
  margin-bottom: 12px;
  margin-top: 0px;
}

.emotion-2 {
  font-size: 14px;
  line-height: 20px;
  font-weight: normal;
  color: #505256;
  margin-top: 0;
  margin-bottom: -40px;
}

.emotion-4 {
  padding: 0;
  margin: 0 -24px;
}

.emotion-3 {
  text-align: center;
  padding: 25px;
}

<div>
  <div>
    <h1
      class="emotion-0"
    >
      Gathering your stats...
    </h1>
    <div
      class="emotion-5"
    >
      <h1
        class="emotion-1"
      >
        Challenges
      </h1>
      <h2
        class="emotion-2"
      >
        Community challenges joined and completed.
      </h2>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-3"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
